### PR TITLE
ENH: use Datafit class

### PR DESCRIPTION
**Pull request by mathurinm**

### CODE REVIEW

**Comment by mathurinm** [comment]
I guess we should maintain Xw and not R (for logreg for example). It may require storing Xj @ y in the datafit class to avoid recomputing it at every gradient step, but I am not sure how to handle it for CV or more generally when new X and y's are passed

**Comment by QB3** [comment]
> I don't know how much more costly it is to do Xj @ (y - Xw) than Xj @ R

I would bet on an additional O(n) per update of coordinate descent.
IMO the 2 questions become:
- how much do we loose with Xj @ (y - Xw) instead of Xj @ R
- is it possible to keep X_j R for quadratic datafit, and Xw for other with too much code duplication

File: andersoncd/solver.py [hunk 1]
```diff
@@ -246,21 +244,22 @@ def solver(
...
         Xj = X[:, j]
         old_w_j = w[j]
         w[j] = penalty.prox_1d(
-            old_w_j + Xj @ R / lc[j], n_samples / lc[j], j)
+            old_w_j + datafit.gradient_scalar(
+                X, w, Xw, j) / lc[j], n_samples / lc[j], j)
+        # TODO use datafit.grad_j and datafit.lipschitz
         if w[j] != old_w_j:
-            R += (old_w_j - w[j]) * Xj
+            Xw -= (old_w_j - w[j]) * Xj
```
Comment (QB3) [inline andersoncd/solver.py:265]: This change leads to overflow, and I do ont understand why

Comment (QB3) [reply andersoncd/solver.py:265]: I think this  is good now

File: andersoncd/datafits.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        pass
+
+    def initialize(self, X, y):
+        self.Xty = X.T @ y  # TODO handle centering
+        self.lipschitz = (X ** 2).sum(axis=0)  # TODO beware of n_samples
+
+    def value(self, X, y, w, Xw):
+        return np.sum((y - Xw) ** 2) / (2 * len(Xw))
+
+    # @staticmethod
```
Comment (mathurinm) [inline andersoncd/datafits.py:23]: rm and below too

File: andersoncd/datafits.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    def gradient(self, X, y, w, Xw):
+        return X.T @ Xw - self.Xty
+
+    # @staticmethod
+    def gradient_scalar(self, X, w, Xw, j):
+        return (X[:, j] @ Xw - self.Xty[j])
```
Comment (mathurinm) [inline andersoncd/datafits.py:29]: rm parenthesis

File: andersoncd/solver.py [hunk 4]
```diff
@@ -1,22 +1,16 @@
...
                 return_n_iter=False, verbose=0,):
     r"""Compute optimization path with Celer primal as inner solver.
 
-    With `n = len(y)` and `p = len(w)` the number of samples and features,
-    the loss is:
-
-    .. math::
-
-        \frac{||y - X w||_2^2}{2 n} + \alpha \sum_1^p penalty(|w_j|)
+    The loss is customised by passing various choices of datafit and penalty.
```
Comment (mathurinm) [inline andersoncd/solver.py:13]: suggestion: the loss is `datafit.value() + penalty.value()`

File: andersoncd/solver.py [hunk 5]
```diff
@@ -106,6 +104,7 @@ def solver_path(X, y, penalty, eps=1e-3, n_alphas=100, alphas=None,
     # X_dense, X_data, X_indices, X_indptr = _sparse_and_dense(X)
 
     if alphas is None:
+        # TODO pass datafit.gradient at 0
```
Comment (mathurinm) [inline andersoncd/solver.py:107]: could this be  done in this PR ?

File: andersoncd/solver.py [hunk 6]
```diff
@@ -137,13 +134,13 @@ def solver_path(X, y, penalty, eps=1e-3, n_alphas=100, alphas=None,
             if coef_init is not None:
                 w = coef_init.copy()
                 p0 = max((w != 0.).sum(), p0)
-                R = y - X @ w
+                Xw = X @ w  # TODO Xw should be used instead
```
Comment (mathurinm) [inline andersoncd/solver.py:137]: This is no longer a TODO, it is done, right ?

File: andersoncd/solver.py [hunk 7]
```diff
@@ -161,22 +158,23 @@ def solver_path(X, y, penalty, eps=1e-3, n_alphas=100, alphas=None,
 
 
 def solver(
-        X, y, penalty, w, R, norms_X_col, max_iter=50,
+        X, y, datafit, penalty, w, Xw, max_iter=50,
         max_epochs=50_000, p0=10, tol=1e-4, use_acc=True, K=5, verbose=0):
     """
-    penalty: Penalty object
+    datafit : instance of Datafit
```
Comment (mathurinm) [inline andersoncd/solver.py:164]: extra space

**Comment by QB3** [comment]
> WDYT of doing it properly and putting the n_samples in the lipschitz constant, in the gradient, etc ? IMO this will be easier for external contributors, and may solve ussome headaches without being more costly

seems better than what we currently have indeed. 
I propose to do it in an other PR

File: andersoncd/solver.py [hunk 8]
```diff
@@ -246,21 +243,23 @@ def solver(
 
 
 @njit
-def _kkt_violation(w, X, R, penalty, ws):
+def _kkt_violation(w, X, Xw, datafit, penalty, ws):
     n_samples = X.shape[0]
     grad = np.zeros(ws.shape[0])
     for idx, j in enumerate(ws):
-        grad[idx] = X[:, j] @ R / n_samples
+        grad[idx] = datafit.gradient_scalar(X, w, Xw, j) / n_samples  # TODO
```
Comment (mathurinm) [inline andersoncd/solver.py:250]: I would remove the n_samples and implement the exact gradient, less error prone for the future IMO

File: andersoncd/solver.py [hunk 9]
```diff
@@ -246,21 +243,23 @@ def solver(
...
-    n_samples = R.shape[0]
+def _cd_epoch(X, w, Xw, datafit, penalty, lc, feats):
+    n_samples = Xw.shape[0]
     for j in feats:
         Xj = X[:, j]
         old_w_j = w[j]
         w[j] = penalty.prox_1d(
-            old_w_j + Xj @ R / lc[j], n_samples / lc[j], j)
+            old_w_j - datafit.gradient_scalar(
+                X, w, Xw, j) / lc[j], n_samples / lc[j], j)
```
Comment (mathurinm) [inline andersoncd/solver.py:262]: to fix here once n_samples is added into lipschitz

File: andersoncd/solver.py (hunk 1 above)
Comment (mathurinm) [reply andersoncd/solver.py:265]: yes :)

File: andersoncd/solver.py (hunk 5 above)
Comment (QB3) [reply andersoncd/solver.py:107]: Is it clear that all penalties do have an alpha_max ?

**Pull request merged**